### ISSUE TITLE
Rule to disable extras matching

### ISSUE THREAD

**Issue by Razzeee** [tree-sitter/tree-sitter#931]
I would like to be able to express, that an extra is not a valid match right now.

```
  extras: ($) => [
    $.block_comment,
    $.line_comment,
    /[\s\uFEFF\u2060\u200B]|\\\r?\n/,
  ],
```

So my extras are comments and spaces right now.

Usually you could use `token.immediate`, but the structure is not shallow enought:
```
negate_expr: ($) =>
      seq(
        alias($.minus_without_trailing_whitespace, $.operator_identifier),
        $._atom
      ), // todo disallow whitespace
```

```
================================================================================
Negate
================================================================================

negate n =
  -n

--------------------------------------------------------------------------------

(file
  (value_declaration
    (function_declaration_left
      (lower_case_identifier)
      (lower_pattern
        (lower_case_identifier)))
    (eq)
    (negate_expr
      (operator_identifier)
      (value_expr
        (value_qid
          (lower_case_identifier))))))
```

I basically want to make sure that `-n` isn't `- n`.

I think I actually do this in my external scanner, but that's not ideal too I think.

```
     // Handle minus without a whitespace for negate
            if (valid_symbols[MINUS_WITHOUT_TRAILING_WHITESPACE])
            {
                if (lexer->lookahead == '-')
                {
                    advance(lexer);
                    auto lookahead = lexer->lookahead;
                    if ((lookahead >= 'a' && lookahead <= 'z') || (lookahead >= 'A' && lookahead <= 'Z') || lookahead == '(')
                    {
                        lexer->result_symbol = MINUS_WITHOUT_TRAILING_WHITESPACE;
                        lexer->mark_end(lexer);

                        return true;
                    }
                    return false;
                }
            }
```
So I basically peek at the next char and decide which kind of a minus it is based on that.

I think it would also be helpful for comments vs strings. Even if prec and sometimes token.immediate seemt to be enough.

**Comment by ahlinc** [comment]
I also faced with this limitation when tried to implement a language grammar by following its specification but when I realized that for an editor it's better to have less restrictive grammar it became not a problem. But if to use tree-sitter as a strict parser it would be good to have such option like local extras for some rule. I suppose that it would be possible to extend tree-sitter's `dsl.js` with an additional `extras(extras_list, $rule)` action that can be applied to any rule what would mean that a node generated by this rule and all its children have the different redefined extras. In this case global extras would became just a special case like the `extras(.... $.startRule)` action applied to starting rule. As I understand such behavior doesn't possible to implement without changing tree-sitter's internal lexer logic.

**Comment by aminya** [comment]
Yeah, this should be definitely added. Currently, the languages write external scanners to support `seq` without `extras` in between. I wish we could a function like `seq_immediate`.

For example, Julia writes an external scanner just to support `immediate_parens` after a function name:

https://github.com/tree-sitter/tree-sitter-julia/blob/12ea597262125fc22fd2e91aa953ac69b19c26ca/src/scanner.c#L15-L21

**Comment by aminya** [comment]
I tried to see how this can be fixed. `token.immediate` is very similar to what we need here with this difference that it creates a single token in the AST, while we want it to have separate parts for the children that make up that token. 

`immediate_token` function is used here:

https://github.com/tree-sitter/tree-sitter/blob/2923c9cb62c964371ed7d6995ca1238356b00b45/cli/src/generate/rules.rs#L102

It calls `add_metadata` function here, which

https://github.com/tree-sitter/tree-sitter/blob/2923c9cb62c964371ed7d6995ca1238356b00b45/cli/src/generate/rules.rs#L434

`is_token` is used inside `extract_tokens_in_rule` and then `extract_token`
https://github.com/tree-sitter/tree-sitter/blob/2923c9cb62c964371ed7d6995ca1238356b00b45/cli/src/generate/prepare_grammar/extract_tokens.rs#L193

https://github.com/tree-sitter/tree-sitter/blob/2923c9cb62c964371ed7d6995ca1238356b00b45/cli/src/generate/prepare_grammar/extract_tokens.rs#L233

So, I think `extract_token` should be modified in some way so it returns the children that make up that token. If @maxbrunsfeld can clarify if this is possible, it would be good.

Otherwise, we just need to continue use external scanners.

**Comment by ahlinc** [comment]
> Currently, the languages write external scanners to support `seq` without `extras` in between. I wish we had a function like `immediate` to show this.

The case with custom extras is more generic that just requirement of immediate rules and it would help to eliminate a bunch of rules duplication.

**Comment by masaeedu** [comment]
Does anyone know of a workaround that allows one to correctly define (in user code) an `immseq` primitive that works equivalently to `seq`, but does not allow spaces in between the sequenced objects? I think this should be possible by recursively traversing the forest represented by the list of arguments, finding "boundary" tokens and peppering in `token.immediate` in the right places. I'm embarking on the project right now, but does anyone know of an example of this already being implemented somewhere (or see reasons why this is impossible)?

**Comment by alaviss** [comment]
Currently in [tree-sitter-nim](https://github.com/alaviss/tree-sitter-nim) I have to use external scanner to test for immediate tokens. In Nim this is not due to strictness but rather for disambiguation:

```nim
a (b) # command call with parenthesized expression
a(b)  # regular call with identifier
1 + 1 # infix expression
1 +1  # command call with prefix expression
```

A problem arise when the external scanner /also/ handles extras: It is not possible to just "reject" an immediate token when an extras (ie. comment) also exists, because that'd mean the extras won't be matched as well.

So instead a state tracking system is used:

- Set "no immediates" flag after scanning an extras
- Remove the flag after immediates are no longer looked for (via issuing a "synchronize" extras)

But an issue arises from this, take the following code:

```nim
foo bar(10)
#  ^ External scanner triggered, read the space and set NoImmediates
#   ^ External scanner triggered again, but immediates are being looked for, so can't remove the flag.
#      ^ External scanner triggered after some progress but the NoImmediates flag is still set here!
```

There are a few "solutions" to this issue:
- Preemptively clear the flag if lookahead can verify that none of the immediate tokens will match. This will require a restructuring of the external scanner and I don't even know if there will be other edge cases.
- Track "progress" via `get_column()`, but this incurs an insane performance hit.

I'd love for disabling extras matching to become an actual feature to avoid the headache of this.

**Comment by acristoffers** [comment]
I'm also in need of a way to disable extras for a rule. I'm writing a MATLAB parser and it if I try do detect escape sequences inside a string with the internal scanner, it ends up detecting comments too, which is wrong. I worked around it by using a scanner, but then I cannot apply the rules for single-quoted strings because there is no way of differentiating a transpose `A'` from a string. The internal-scanner rules I have work just fine, as long as I don't have the comments rule in extras, but of course that breaks other things, so disabling extras for just my strings rule would be ideal.
